New Style Button Command and New Alpha Button Command

DoctorRyan opened the pull request:
This is an update to the palette viewer that adds a more convenient plus button for adding colors and a more clear Color 0 button to avoid people confusing this with white color.  The original author was Turtletooth.  Thanks to him for writing the code.

RodneyBaker commented:
@shun-iwasawa  As there are several changes to the palette window included in this PR outside of adding the new style button and changing color_0 visual appearance it is appropriate to get your feedback on how this might effect current users, especially those in Japanese studios.

shun-iwasawa commented:
@DoctorRyan 
> The original author was Turtletooth. 

First of all, please let me clarify, did you get a permission from him to use his code or is it released under certain license?

RodneyBaker commented:
Shun,
@turtletooth (Jeremy Bullock) has stated this code is under the same BSD license as Opentoonz.

turtletooth commented:
I don't mind if anything from Tahoma gets back-ported to OpenToonz.

turtletooth commented:
However, I would also bring in my change to the colors of the checkerboard to make the colors more in line with other software.
![colors1](https://user-images.githubusercontent.com/4576381/86814205-51bbf800-c03e-11ea-8bf1-9527bc49eef0.PNG)

turtletooth commented on toonz/sources/include/toonzqt/paletteviewergui.h:
```diff
@@ -116,6 +116,9 @@ class DVAPI PageViewer final : public QFrame, public TSelection::View {
   NameDisplayMode getNameDisplayMode() const { return m_nameDisplayMode; }
   void setNameDisplayMode(NameDisplayMode mode);
 
+  void toggleShowStyleIndex();
```
I'm leaving comments for Ryan:
This makes the style indexes in the corner toggle-able.  I'm not sure that is desired for OpenToonz.

turtletooth commented on toonz/sources/toonzqt/paletteviewer.cpp:
```diff
@@ -105,7 +105,7 @@ PaletteViewer::PaletteViewer(QWidget *parent, PaletteViewType viewType,
   m_pageViewerScrollArea = new QScrollArea();
   m_pageViewerScrollArea->setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
-  m_pageViewerScrollArea->setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOn);
+  m_pageViewerScrollArea->setVerticalScrollBarPolicy(Qt::ScrollBarAsNeeded);
```
Most of the changes in this file are beyond the scope of this PR.

turtletooth commented on toonz/sources/toonzqt/paletteviewergui.cpp:
```diff
@@ -41,6 +41,8 @@
 TEnv::IntVar EnvSoftwareCurrentFontSize_StyleName(
     "SoftwareCurrentFontSize_StyleName", 11);
 
+TEnv::IntVar ShowStyleIndex("ShowStyleIndex", 0);
```
Unnecessary if the indexes are going to be permanent.

turtletooth commented on toonz/sources/toonzqt/paletteviewergui.cpp:
```diff
@@ -138,16 +140,16 @@ PageViewer::PageViewer(QWidget *parent, PaletteViewType viewType,
   switch (m_viewType) {
   case LEVEL_PALETTE:
-    defaultChipSize = LargeChips;
+    defaultChipSize = SmallChips;
```
I'd ask before changing the default chip sizes.

turtletooth commented on toonz/sources/toonzqt/paletteviewergui.cpp:
```diff
@@ -226,6 +228,20 @@ int PageViewer::getCurrentStyleIndex() const {
 
+//-----------------------------------------------------------------------------
+
+void PageViewer::toggleShowStyleIndex() {
```
Again unnecessary if keeping the style indexes.

turtletooth commented on toonz/sources/toonzqt/paletteviewergui.cpp:
```diff
@@ -715,10 +750,12 @@ void PageViewer::paintEvent(QPaintEvent *e) {
-            bottomRect.adjust(0, bottomRect.height() - 12, 0, 0);
-          } else
-            bottomRect.adjust(0, bottomRect.height() - 6, 0, 0);
+          if (styleIndex != 0) {
```
I'd remove this change.

turtletooth commented on toonz/sources/toonzqt/paletteviewergui.cpp:
```diff
@@ -766,10 +803,7 @@ void PageViewer::paintEvent(QPaintEvent *e) {
-      QFont tmpFont = p.font();
-
-      if (m_viewMode != SmallChips) {
+      if (m_viewMode != SmallChips && styleIndex != 0) {
```
Remove this change too.

turtletooth commented on toonz/sources/toonzqt/paletteviewergui.cpp:
```diff
@@ -825,13 +858,15 @@ void PageViewer::paintEvent(QPaintEvent *e) {
-      p.setBrush(Qt::white);
-      p.drawRect(indexRect);
-      p.drawText(indexRect, Qt::AlignCenter, QString().setNum(styleIndex));
+      if (ShowStyleIndex == 1) {
```
This can be removed too.

turtletooth commented on toonz/sources/toonzqt/paletteviewergui.cpp:
```diff
@@ -653,6 +671,23 @@ void PageViewer::paintEvent(QPaintEvent *e) {
       // toggle link
       drawToggleLink(p, chipRect, m_page->getStyle(i));
     }
+    if (!m_page->getPalette()->isLocked()) {
```
This is where the new style button is drawn in list view.  Keep this.

turtletooth commented on toonz/sources/toonzqt/paletteviewergui.cpp:
```diff
@@ -880,6 +915,24 @@ void PageViewer::paintEvent(QPaintEvent *e) {
       // draw link indicator
       drawToggleLink(p, chipRect, style);
     }
+    // draw new style chip
```
This is where the new style button is drawn.  Keep this.

turtletooth commented on toonz/sources/toonzqt/paletteviewergui.cpp:
```diff
@@ -955,17 +1008,20 @@ void PageViewer::mousePressEvent(QMouseEvent *event) {
-    m_styleSelection->select(pageIndex);
-    m_styleSelection->makeCurrent();
-
+    if (indexInPage == getChipCount() && !m_page->getPalette()->isLocked()) {
```
This is the logic for pressing the new button.  Keep this.

turtletooth commented on toonz/sources/toonzqt/paletteviewergui.cpp:
```diff
@@ -1325,15 +1382,24 @@ bool PageViewer::event(QEvent *e) {
-        if (shortcutKey > 0)
-          toolTip += QString::fromStdWString(std::wstring(L" (") +
-                                             (wchar_t)shortcutKey + L")");
+        if (styleIndex == 0) {
```
I'd remove these changes.

turtletooth commented on toonz/sources/toonzqt/paletteviewergui.cpp:
```diff
@@ -1325,15 +1382,24 @@ bool PageViewer::event(QEvent *e) {
+        }
       }
     }
+    if (indexInPage == m_page->getStyleCount()) {
```
Keep this.

DoctorRyan commented:
Thanks for the review Turtle.  I apologize for posting so late.  Mr. @shun-iwasawa   As they have stated I was given his blessing to try to import over some of his changes to Opentoonz.  Still some work to be done that I will be sure to get done.  My goal is to try and help make OT more user friendly if possible.  Any suggestions or feedback is most appreciated.  A pleasure to finally meet you.

manongjohn commented:
Build errors that need to be addressed:

C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(133): error C2065: 'm_newPageToolbar': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(134): error C2065: 'm_newPageToolbar': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(134): error C2227: left of '->clear' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(134): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(135): error C2065: 'm_newPageToolbar': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(135): error C2227: left of '->setMovable' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(135): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(136): error C2065: 'm_newPageToolbar': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(136): error C2227: left of '->setIconSize' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(136): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(137): error C2065: 'm_newPageToolbar': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(137): error C2227: left of '->setLayoutDirection' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(137): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(142): error C2065: 'm_newPageToolbar': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(145): error C2065: 'm_newPageToolbar': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(145): error C2227: left of '->addAction' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(145): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(151): error C2065: 'm_newPageToolbar': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(295): error C3861: 'updatePaletteMenu': identifier not found [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(442): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(443): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(443): error C2227: left of '->setLayoutDirection' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(443): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(445): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(451): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(456): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(456): error C2227: left of '->addAction' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(456): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(464): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(464): error C2227: left of '->addSeparator' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(464): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(476): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(476): error C2227: left of '->addAction' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(476): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(489): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(489): error C2227: left of '->addSeparator' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(489): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(490): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(490): error C2227: left of '->addAction' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(490): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(493): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(493): error C2227: left of '->addSeparator' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(493): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(499): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(499): error C2227: left of '->addAction' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(499): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(505): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(505): error C2227: left of '->addAction' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(505): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(508): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(513): error C3861: 'updatePaletteMenu': identifier not found [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(568): error C2065: 'm_newPageToolbar': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(568): error C2227: left of '->actions' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(568): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(605): error C2039: 'updatePaletteMenu': is not a member of 'PaletteViewer' [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\include\toonzqt/studiopaletteviewer.h(30): note: see declaration of 'PaletteViewer'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(606): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(608): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(608): error C2227: left of '->actions' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(608): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(609): error C3861: 'getPalette': identifier not found [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(610): error C2065: 'm_pageViewer': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(610): error C2227: left of '->getViewMode' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(610): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(614): error C3861: 'tr': identifier not found [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(615): error C3861: 'tr': identifier not found [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(616): error C3861: 'tr': identifier not found

turtletooth commented:
@shun-iwasawa I updated the license file at Tahoma:
https://github.com/turtletooth/tahoma/blob/master/LICENSE.txt

DoctorRyan commented:
Sorry, my commit history looks like a train wreck.  However, after breaking things 10345 times I am back on track with the commit.  Removed all unnecessary lines of code, fixed changes to UI that didn't need to be changed as well.  Tested on my local and looks good so far.  I will take care of the checkboard pattern to make it not so garish next.

DoctorRyan commented:
@manongjohn Did you take a look and see if that cleared up the issues for you?
